### PR TITLE
Deal with exception raised while parsing

### PR DESCRIPTION
**Pull request by alexsnaps**

### CODE REVIEW

**Comment by alexsnaps** [comment]
Sorry, looks like this broke something

File: ext/mruby_engine/mruby_engine.c [hunk 1]
```diff
@@ -1,13 +1,15 @@
 #include "mruby_engine_private.h"
 #include "platform.h"
 #include <mruby.h>
+#include <stdio.h>
```
Comment (alanly) [inline ext/mruby_engine/mruby_engine.c]: Is this used anywhere?

File: ext/mruby_engine/mruby_engine.c [hunk 2]
```diff
@@ -209,10 +210,22 @@ static struct RProc *generate_code(
...
   struct mrb_parser_state *parser_state = mrb_parser_new(state);
   parser_state->s = source;
   parser_state->send = source + strlen(source);
-  mrb_parser_parse(parser_state, context);
+
+  MRB_TRY(state->jmp) {
+    mrb_parser_parse(parser_state, context);
+  }
+  MRB_CATCH(state->jmp) {
+    parser_state->nerr = 1;
```
Comment (sgnr) [inline ext/mruby_engine/mruby_engine.c]: Wait so this was correct after all? I understood the opposite.

Comment (sgnr) [inline ext/mruby_engine/mruby_engine.c]: Shouldn't we generate an error with `me_host_syntax_error_new` from `state->exc`?

Comment (alexsnaps) [inline ext/mruby_engine/mruby_engine.c]: It is... the `parser_state` is fine... And that way we get the error as we already got to it

Comment (sgnr) [inline ext/mruby_engine/mruby_engine.c]: I have a hard time believing `parser_state->error_buffer[0]` will be set in every situation where we would raise here. If so isn't the problem that the underlying code sets `parser_state->error_buffer[0]`, but no `parser_state->nerr`?

Comment (sgnr) [inline ext/mruby_engine/mruby_engine.c]: It makes no sense to me to set `error_buffer`, but not `nerr`.

Comment (alexsnaps) [inline ext/mruby_engine/mruby_engine.c]: Got you! I think that's much better now, wdyt?

**Comment by sgnr** [comment]
Looks nice. What do you think of declaring `proc = NULL` at the top of the function, having
```c
done:
  mrb_parser_free(parser_state);
  return proc;
```
and using `goto done;` on error?

**Comment by sgnr** [comment]
Feel free to ignore my suggestion if you do not think it improves the code.

**Comment by clayton-shopify** [comment]
This fixes https://hackerone.com/reports/180695.

**Pull request merged**